src/transformers/models/clip/modeling_tf_clip.py

```diff
@@ -0,0 +1,1528 @@
+            :class:`~transformers.TFCLIPVisionModel`.
+        text_model_output(:obj:`TFBaseModelOutputWithPooling`):
+            The output of the :class:`~transformers.TFCLIPTextModel`.
+        vision_model_output(:obj:`TFBaseModelOutputWithPooling`):
+            The output of the :class:`~transformers.TFCLIPVisionModel`.
+    """
+
+    loss: Optional[tf.Tensor] = None
+    logits_per_image: tf.Tensor = None
+    logits_per_text: tf.Tensor = None
+    text_embeds: tf.Tensor = None
+    image_embeds: tf.Tensor = None
+    text_model_output: TFBaseModelOutputWithPooling = None
+    vision_model_output: TFBaseModelOutputWithPooling = None
+
+    def to_tuple(self) -> Tuple[Any]:
+        return tuple(
+            self[k] if k not in ["text_model_output", "vision_model_output"] else getattr(self, k).to_tuple()
+            for k in self.keys()
+        )
+
+
+class TFCLIPVisionEmbeddings(tf.keras.layers.Layer):
+    def __init__(self, config: CLIPVisionConfig, **kwargs):
+        super().__init__(**kwargs)
+
+        self.embed_dim = config.hidden_size
+        self.image_size = config.image_size
+        self.patch_size = config.patch_size
+
+        self.num_patches = (self.image_size // self.patch_size) ** 2
+        self.num_positions = self.num_patches + 1
+
+        self.config = config
+
+        self.patch_embedding = tf.keras.layers.Conv2D(
+            filters=self.embed_dim,
+            kernel_size=self.patch_size,
+            strides=self.patch_size,
+            padding="valid",
+            data_format="channels_last",
+            use_bias=False,
+            kernel_initializer=get_initializer(self.config.initializer_range),
+            name="patch_embedding",
+        )
+
+    def build(self, input_shape: tf.TensorShape):
+
+        # Q: should we use normal initializer?
+        # initializer = tf.random_normal_initializer(mean=0.0, stddev=1.0)
+        initializer = get_initializer(self.config.initializer_range)
+
+        self.class_embedding = self.add_weight(
+            shape=(self.embed_dim,), initializer=initializer, trainable=True, name="class_embedding"
+        )
+
+        with tf.name_scope("position_embedding"):
+            self.position_embedding = self.add_weight(
+                shape=(self.num_positions, self.embed_dim),
+                initializer=get_initializer(self.config.initializer_range),
+                trainable=True,
+                name="embeddings",
+            )
+
+        super().build(input_shape)
+
+    def call(self, pixel_values: tf.Tensor) -> tf.Tensor:
+
+        batch_size, num_channels, height, width = shape_list(pixel_values)
+
+        # When running on CPU, `tf.nn.conv2d` doesn't support `NCHW` format.
+        # So change the input format from `NCHW` to `NHWC`.
+        # shape = (batch_size, in_height, in_width, in_channels=num_channels)
+        pixel_values = tf.transpose(pixel_values, perm=(0, 2, 3, 1))
```

Comment:
Should we properly document we expect inputs in the NCHW format? I think TF users might be more used to NHWC?

Reply:
@ydshieh I discussed this with Sylvain, and we decided keeping NCHW is definitely better!

Reply:
`CLIP_VISION_INPUTS_DOCSTRING` has already mentioned

```
pixel_values: ... and each example must have the shape :obj:`(batch_size, num_channels, height, width)`):
```
I just added (for `TFCLIPVisionEmbeddings`)
```
  def call(self, pixel_values: tf.Tensor) -> tf.Tensor:
      """`pixel_values` is expected to be of NCHW format."""
```